shutdown.8
.\"{{{}}}
.\"{{{  Title
.TH SHUTDOWN 8 "May 13, 1993" "" "Linux Programmer's Manual"
.\"}}}
.\"{{{  Name
.SH NAME
shutdown \- bring the system down
.\"}}}
.\"{{{  Synopsis
.SH SYNOPSIS
.B /etc/shutdown
.RB [ \-t
.IR sec ]
.RB [ \-rkhncf ]
.I time
.RI [ warning-message ]
.\"}}}
.\"{{{  Description
.SH DESCRIPTION
.B Shutdown
brings the system down in a secure way.  All logged in users are
notified that the system is going down, and \fIlogin(1)\fP is blocked.
It is possible to shut the system down immideately, or after a delay.
All processes are first notified that the system is going down by the
signal \s-2SIGTERM\s0.  This gives programs like \fIvi(1)\fP the time to
save the file being editted, mail and news processing programs a chance
to exit cleanly, etc.  When the system is idle, the program
\fB/etc/brc\fP is run, if present.  Then the system is either rebooted
by \fIreboot(1)\fP or halted by \fIhalt(1)\fP.
.\"}}}
.\"{{{  Options
.SH OPTIONS
.\"{{{  -t sec
.IP "\fB\-t\fP \fIsec\fP"
Wait \fIsec\fP seconds between sending processes the warning and
the kill signal.
.\"}}}
.\"{{{  -k
.IP \fB\-k\fP
Don't really shutdown; only send the warning messages to everybody.
.\"}}}
.\"{{{  -r
.IP \fB\-r\fP
Reboot after shutdown.
.\"}}}
.\"{{{  -h
.IP \fB\-h\fP
Halt after shutdown.
.\"}}}
.\"{{{  -n
.IP \fB\-n\fP
Don't sync before rebooting or halting.
.\"}}}
.\"{{{  -f
.IP \fB\-f\fP
Do a `fast' reboot, that means no filesystems will be checked.
.\"}}}
.\"{{{  -c
.IP \fB\-c\fP
Cancel an already running shutdown. With this option it is ofcourse
not possible to give the \fBtime\fP argument, but you can enter a
explanitory message on the command line that will be sent to all
users.
.\"}}}
.\"{{{  time
.IP \fItime\fP
When to shutdown.
.\"}}}
.\"{{{  warning-message
.IP \fIwarning-message\fP
Message to send to all users.
.\"}}}
.PP
The \fItime\fP argument can have different formats.  First, it can be an
absolute time in the format \fIhh:mm\fP, in which \fIhh\fP is the hour
(1 or 2 digits) and \fImm\fP is the minute of the hour (in two digits).
Second, it can be in the format \fB+\fP\fIm\fP, in which \fIm\fP is the
number of minutes to wait.  The word \fBnow\fP is an alias for \fB+0\fP.
.PP
The \fB\-f\fP flag means `reboot fast'.  This only creates an advisory
file \fB/etc/fastboot\fP which can be tested by the system when it comes
up again.  The boot rc file can test if this file is present, and decide
not to run \fIfsck(1)\fP since the system has been shut down in the
proper way.  After that, the boot process should remove
\fB/etc/fastboot\fP.
.\"}}}
.\"{{{  Files
.SH FILES
.nf
/etc/fastboot
/etc/halt
/etc/reboot
.fi
.\"}}}
.\"{{{  Author
.SH AUTHOR
Miquel van Smoorenburg, miquels@drinkel.nl.mugnet.org
.\"}}}
.\"{{{  See also
.SH "SEE ALSO"
fsck(1), init(1), halt(8), reboot(8)
.\"}}}


History
 Version:	1.0, 01-Feb-92
		- Initial version, very primitive for the Minix
		  operating system. Required some mods. to the
		  kernel.

 		1.1, 30-Apr-92
		- Read manual wrong: there is no 'action' field called
		  process, but all entries are of type process. Every
		  'process' get exec'ed by /bin/sh -c 'exec command'.
		- Rapidly respawning processes are caught in the act.
		- _SYSV support is really Linux support,
		  done by poe@daimi.aau.dk on 25-Mar-92.

		1.2, 16-Jun-92
		- Bugreport from Michael Haardt ; removed deadlock
		  and added 'waitpid' instead of 'wait' for SYSV.

		1.3, 05-Jul-92
		- Got a 386, so installed Linux. Added 'soft' reboot
		  to be default under linux. Fixed some typos.

		2.0 08-Dec-92
		- Rewrote the code totally, so started with a new
		  version number.
		- Dropped Minix support, this code now is Linux - specific.
		- With TEST switch on, this init & telinit can
		  run standalone for test purposes.

		2.1 28-Jan-93
		- Fixed a bug with 'boot' and 'once'.
		- Check 'initdefault' for validity.
		- Reckognizes "single" as command line argument.
		- Retries execvp with 'sh -c exec ..' if command
		  is a shell script. (shouldn't execvp do this?)
		- Added patches to use syslog if defined.

		2.2 02-Mar-93
		- Made wait()'s asynchronous
		- Changed whole thing to one big state machine
		- Now using 'pseudo levels' # & * for SYSINIT & BOOT
		- Added a new type of 'action', called leave. This
		  process will be executed when the system goes from a
		  runlevel specified in it's runlevel field to a 
		  level that's not. Nice to bring down NFS and the like.

		2.3 24-Mar-93
		- Ripped out the 'leave' action. To difficult, and
		  unneeded.
		- Going single user now kills _all_ processes.
		- Added '-t secs' option to all commands.
		- This version is stable enough to post.

		2.4a 22-Apr-93
		- Fixed last to reckognize BSD style wtmp logging.
		- Changed init to write wtmp records that are
		  SysV compliant but are also reckognized by the
		  BSD last. Added a '+' option to the 'process'
		  field of inittab, for getties that want to do
		  their own utmp/wtmp housekeeping (kludge!).
		- Now accepts a runlevel on the command line,
		  and reckognizes the 'auto' argument. (Sets the
		  environment variable AUTOBOOT to YES)

		2.4g 15-May-93
		- Changed init to really catch SIGPWR 'cause we
		  hooked up an UPS to the Linux machine. The
		  keyword for catching the TreeFingerSalute is
		  now "ctrlaltdel" instead of "power{wait,fail}".

		2.4  24-May-93
		- Send out the official version into the world as
		  SysVinit-2.4.tar.z.


shutdown.c
/*
 * shutdown.c	- Shut down the system.
 *
 * Usage:	  shutdown [-krhnf] time [warning message]
 *		  -k: don't really shutdown, only warn.
 *		  -r: reboot after shutdown.
 *		  -h: halt after shutdown.
 *		  -n: don't sync before reboot or halt.
 *		  -f: do a 'fast' reboot.
 *		  -c: cancel an already running shutdown.
 *		  -t secs: delay between SIGTERM and SIGKILL
 *
 * Author:	  Miquel van Smoorenburg, miquels@drinkel.nl.mugnet.org
 *
 * Version:	- 0.01 Long time ago
 *		- 1.1  07-12-1992
 *		- 1.2  15-12-1992
 *		- 1.3  24-03-1993
 *		- 1.4  14-05-1993
 */
#include <sys/types.h>
#include <sys/stat.h>
#include <time.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h> 
#include <signal.h>
#include <fcntl.h>

char *Version = "@(#) shutdown 1.4 14-05-1993 MvS";

/* Some pathnames */
#define NOLOGIN		"/etc/nologin"
#define FASTBOOT	"/etc/fastboot"
#define INIT		"/etc/init"
#define REBOOT		"/etc/reboot"
#define HALT		"/etc/halt"
#define SDPID		"/etc/shutdownpid"

int dontshut = 0;	/* Don't shutdown, only warn	*/
int do_reboot = 0;	/* Reboot after shutdown	*/
int halt = 0;		/* Halt after shutdown		*/
int dosync = 1;		/* Sync before reboot or halt	*/
int quick = 0;		/* Don't send processes SIGTERM */
int fastboot = 0;	/* Do a 'fast' reboot		*/
char reason[256];	/* Warning message		*/
char *sltime = 0;	/* Sleep time			*/

/* From "wall.c" */
extern void wall();

/*
 * Break off an already running shutdown.
 */
void stopit()
{
  unlink(NOLOGIN);
  unlink(FASTBOOT);
  unlink(SDPID);
  printf("\nShutdown cancelled.\n");
  exit(0);
}

/*
 * Show usage message.
 */
void usage()
{
 fprintf(stderr, "%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n",
	"Usage:\t  shutdown [-krhnf] time [warning message]",
	"\t\t  -k:      don't really shutdown, only warn.",
	"\t\t  -r:      reboot after shutdown.",
	"\t\t  -h:      halt after shutdown.",
	"\t\t  -n:      don't sync before reboot or halt.",
	"\t\t  -f:      do a 'fast' reboot.",
	"\t\t  -c:      cancel a running shutdown.",
	"\t\t  -t secs: delay between warning and kill signal.");
  exit(1);
}

/*
 * Tell everyone the system is going down in 'mins' minutes.
 */
void warn(mins)
int mins;
{
  char buf[512];
  int len;

  strcpy(buf, reason);
  len = strlen(buf);

  if (mins == 0)
	sprintf(buf + len, "\rThe system is going down NOW !!\r\n");
  else
  	sprintf(buf + len,
		"\rThe system is going DOWN in %d minutes !!\r\n", mins);
  wall(buf);
}

/*
 * Create the /etc/nologin file.
 */
void donologin()
{
  FILE *fp;
  time_t t;

  time(&t);
  t += 300;

  unlink(NOLOGIN);
  if ((fp = fopen(NOLOGIN, "w")) != NULL) {
  	fprintf(fp, "\rThe system is going down on %s\r\n", ctime(&t));
  	if (reason[0]) fputs(reason, fp);
  	fclose(fp);
  }
}

/*
 * Execute /etc/reboot or /etc/halt.
 */
void shutdown()
{
  char *prog = (char *)0;
  char *args[6];
  int argp = 0;

  warn(0);
  sleep(2); /* Give telnetd etc the chance to show this message */
  if (dontshut) stopit();

  if (do_reboot) prog = REBOOT;
  if (halt) prog = HALT;

  args[argp++] = prog;
  if (!dosync) args[argp++] = "-n";
  if (quick)   args[argp++] = "-q";
  if (sltime) {
	args[argp++] = "-t";
	args[argp++] = sltime;
  }
  args[argp]   = (char *)NULL;

  unlink(NOLOGIN);
  unlink(SDPID);

  if (prog) {
  	execv(prog, args);
  	fprintf(stderr, "shutdown: cannot execute %s\n", prog);
	unlink("/etc/fastboot");
	exit(1);
  }
  if (dosync) sync();
  if (sltime)
	execl(INIT, "telinit", "-t", sltime, "s", NULL);
  else
	execl(INIT, "telinit", "s", NULL);
  fprintf(stderr, "shutdown: cannot get into single user mode\n");
  unlink(FASTBOOT);	
  exit(1);
}

/*
 * Main program.
 * Process the options and do the final countdown.
 */
int main(argc, argv)
int argc;
char **argv;
{
  extern int getopt();
  extern int optind; 
  int c, wt, hours, mins;
  struct tm *lt;
  time_t t;
  char *n;
  int didnolog = 0;
  int cancel = 0;
  int pid = 0;
  FILE *fp;

  /* We can be installed setuid root (executable for a special group) */
  setuid(geteuid());

  if (getuid() != 0) {
  	fprintf(stderr, "shutdown: must be root.\n");
  	exit(1);
  }

  while((c = getopt(argc, argv, "cqkrhnft:")) != EOF) {
  	switch(c) {
		case 'c': /* Cancel an already running shutdown. */
			cancel = 1;
			break;
		case 'q': /* Do a quick reboot */
			quick = 1;
			break;
  		case 'k': /* Don't really shutdown, only warn.*/
  			dontshut = 1;
  			break;
  		case 'r': /* Automatic reboot */
  			do_reboot = 1;
  			halt = 0;
  			break;
  		case 'h': /* Halt after shutdown */
  			halt = 1;
  			do_reboot = 0;
  			break;
  		case 'n': /* Don't sync after shutdown */
  			dosync = 0;
  			break;
  		case 'f': /* Don't perform fsck after next boot */
  			fastboot = 1;
  			break;
		case 't': /* Delay between TERM and KILL */
			sltime = optarg;
			break;
  		default:
  			usage();
  			break;	
  	}
  }
  /* Read pid of running shutdown from a file */
  if ((fp = fopen(SDPID, "r")) != NULL) {
	fscanf(fp, "%d", &pid);
	fclose(fp);
  }

  /* Read remaining words */
  reason[0] = 0;
  for(c = optind + (cancel == 0); c < argc; c++) {
  	strcat(reason, argv[c]);
  	strcat(reason, " ");
  }
  if (reason[0]) strcat(reason, "\r\n");

  /* See if we want to run or cancel. */
  if (cancel) {
	if (pid <= 0) {
		fprintf(stderr, "shutdown: cannot find pid of running shutdown.\n");
		exit(1);
	}
	if (kill(pid, SIGINT) < 0) {
		fprintf(stderr, "shutdown: not running.\n");
		exit(1);
	}
	if (reason[0]) wall(reason);
	exit(0);
  }
  
  /* Check syntax. */
  if (optind == argc) usage();
  n = argv[optind++];
  
  /* See if we are already running. */
  if (pid > 0 && kill(pid, 0) == 0) {
	fprintf(stderr, "shutdown: already running.\n");
	exit(1);
  }

  /* Create a new PID file. */
  unlink(SDPID);
  umask(022);
  if ((fp = fopen(SDPID, "w")) != NULL) {
	fprintf(fp, "%d", getpid());
	fclose(fp);
  } else
	fprintf(stderr, "shutdown: warning: cannot open %s\n", SDPID);

  for(c = 1; c < _NSIG; c++) signal(c, SIG_IGN);
  signal(SIGINT, stopit);

  /* Go to the root directory */
  chdir("/");
  if (fastboot) close(open(FASTBOOT, O_CREAT | O_RDWR, 0644));

  if (!strcmp(n, "now") || !strcmp(n, "+0")) shutdown();

  if (n[0] == '+') {
  	wt = atoi(n + 1);
  	if (wt == 0) usage();
  } else {
  	if (sscanf(n, "%d:%2d", &hours, &mins) != 2) usage();
  	if (hours > 23 || mins > 59) usage();
  	time(&t);
  	lt = localtime(&t);
  	wt = (60*hours + mins) - (60*lt->tm_hour + lt->tm_min);
  	if (wt < 0) wt += 1440;
  }
  if (wt < 15 && wt!= 10 && wt != 5 && wt != 1) warn(wt);

  while(wt) {
  	if (wt <= 5 && !didnolog) {
  		donologin();
  		didnolog++;
  	}
  	if (wt == 15 || wt == 10 || wt == 5 || wt == 1) warn(wt);
  	sleep(60);
  	wt--;
  }
  shutdown();
  return(0); /* Never happens */
}
